streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1432,7 +1480,9 @@ public void cleanUp() {
         validateIsRunningOrRebalancing();
         final Set<ThreadMetadata> threadMetadata = new HashSet<>();
         for (final StreamThread thread : threads) {
-            threadMetadata.add(thread.threadMetadata());
+            if (thread.state() != StreamThread.State.DEAD) {
+                threadMetadata.add(thread.threadMetadata());
+            }
         }
         return threadMetadata;
```

Comment:
Do we not need to synchronize this block on the `changeThreadCount` to at least guarantee that the state is correct when the method returns? Otherwise between the `if` and `threadMetadata.add(thread.threadMetadata())` the stream thread might transit to `DEAD`.

Reply:
`changeThreadCount ` would not take care of that we need to sync on the thread state

Reply:
Ah yes, you are right! The stream thread might be replaced or just normally shut down which would not be synchronized on the `changeThreadCount` lock. However, we still do not guarantee that the state is correct when the method returns, because the state could change after the lock is released but before the method returns. At this point with or without lock it doesn't matter. Either we find something that synchronizes the whole method or we can also remove the synchronisation on the stream thread state. And also if we find something that synchronizes the whole method, I am not sure if this guarantee is worth the hassle. WDYT?

Reply:
since changing the thread metadata uses a copy once its is added to the list as not dead it insures that we won't return a dead thread. And after we and it to the list we don't need to care until method is called again